main.tf

```diff
@@ -1,159 +1,198 @@
+      Project     = var.project,
+      Environment = var.environment
+    },
+    var.tags
+  )
 }
 
 resource "aws_subnet" "public" {
-  count = "${length(var.public_subnet_cidr_blocks)}"
+  count = length(var.public_subnet_cidr_blocks)
 
-  vpc_id                  = "${aws_vpc.default.id}"
-  cidr_block              = "${var.public_subnet_cidr_blocks[count.index]}"
-  availability_zone       = "${var.availability_zones[count.index]}"
+  vpc_id                  = aws_vpc.default.id
+  cidr_block              = var.public_subnet_cidr_blocks[count.index]
+  availability_zone       = var.availability_zones[count.index]
   map_public_ip_on_launch = true
 
-  tags {
-    Name        = "PublicSubnet"
-    Project     = "${var.project}"
-    Environment = "${var.environment}"
-  }
+  tags = merge(
+    {
+      Name        = "PublicSubnet",
+      Project     = var.project,
+      Environment = var.environment
+    },
+    var.tags
+  )
 }
 
 resource "aws_route_table_association" "private" {
-  count = "${length(var.private_subnet_cidr_blocks)}"
+  count = length(var.private_subnet_cidr_blocks)
 
-  subnet_id      = "${aws_subnet.private.*.id[count.index]}"
-  route_table_id = "${aws_route_table.private.*.id[count.index]}"
+  subnet_id      = aws_subnet.private[count.index].id
+  route_table_id = aws_route_table.private[count.index].id
 }
 
 resource "aws_route_table_association" "public" {
-  count = "${length(var.public_subnet_cidr_blocks)}"
+  count = length(var.public_subnet_cidr_blocks)
 
-  subnet_id      = "${aws_subnet.public.*.id[count.index]}"
-  route_table_id = "${aws_route_table.public.id}"
+  subnet_id      = aws_subnet.public[count.index].id
+  route_table_id = aws_route_table.public.id
 }
 
 resource "aws_vpc_endpoint" "s3" {
-  vpc_id          = "${aws_vpc.default.id}"
-  service_name    = "com.amazonaws.${var.region}.s3"
-  route_table_ids = ["${aws_route_table.public.id}", "${aws_route_table.private.*.id}"]
+  vpc_id       = aws_vpc.default.id
+  service_name = "com.amazonaws.${var.region}.s3"
+  route_table_ids = flatten([
+    aws_route_table.public.id,
+    aws_route_table.private.*.id
+  ])
+
+  tags = merge(
+    {
+      Name        = "endpointS3",
+      Project     = var.project,
+      Environment = var.environment
+    },
+    var.tags
+  )
 }
 
 #
 # NAT resources
 #
-
 resource "aws_eip" "nat" {
-  count = "${length(var.public_subnet_cidr_blocks)}"
+  count = length(var.public_subnet_cidr_blocks)
 
   vpc = true
 }
 
 resource "aws_nat_gateway" "default" {
-  count = "${length(var.public_subnet_cidr_blocks)}"
+  depends_on = ["aws_internet_gateway.default"]
 
-  allocation_id = "${aws_eip.nat.*.id[count.index]}"
-  subnet_id     = "${aws_subnet.public.*.id[count.index]}"
+  count = length(var.public_subnet_cidr_blocks)
 
-  depends_on = ["aws_internet_gateway.default"]
+  allocation_id = aws_eip.nat[count.index].id
+  subnet_id     = aws_subnet.public[count.index].id
+
+  tags = merge(
```

Comment:
I spun this up, but for some reason, these tags aren't sticking:

```bash
. . .
  # module.vpc.aws_nat_gateway.default[1] will be created
  + resource "aws_nat_gateway" "default" {
      + allocation_id        = (known after apply)
      + id                   = (known after apply)
      + network_interface_id = (known after apply)
      + private_ip           = (known after apply)
      + public_ip            = (known after apply)
      + subnet_id            = (known after apply)
      + tags                 = {
          + "Environment" = "Staging"
          + "Name"        = "gwNAT"
          + "Project"     = "Dinghy"
        }
    }
. . .
```

![image](https://user-images.githubusercontent.com/1774125/71289243-add95c00-233a-11ea-8270-f480dcfd1f33.png)

Reply:
This comment is associated with the NAT gateway resource, but the screenshot is of an internet gateway resource. Looks like we weren't setting tags on the IGW, but the Terraform resource now supports them. Added tags in 4e31487d43f98bd5b8b4370b2529fd346756b7c1.